lib/plugin/instance.rb

```diff
@@ -1257,6 +1257,72 @@ def register_search_group_query_callback(callback)
     DiscoursePluginRegistry.register_search_groups_set_query_callback(callback, self)
   end
 
+  # This is an experimental API and may be changed or removed in the future without deprecation.
+  #
+  # Adds a custom rate limiter to the request rate limiters stack. Only one rate limiter is used per request and the
+  # first rate limiter in the stack that is active is used. By default the rate limiters stack contains the following
+  # rate limiters:
+  #
+  #   `RequestTracker::RateLimiters::User` - Rate limits authenticated requests based on the user's id
+  #   `RequestTracker::RateLimiters::IP` - Rate limits requests based on the IP address
+  #
+  # @param identifier [Symbol] A unique identifier for the rate limiter.
+  #
+  # @param key [Proc] A lambda/proc that defines the `rate_limit_key`.
+  #   - Receives `request` (An instance of `Rack::Request`) as argument.
+  #   - Should return a string representing the rate limit key.
+  #
+  # @param activate_when [Proc] A lambda/proc that defines when the rate limiter should be used for a request.
+  #   - Receives `request` (An instance of `Rack::Request`) as argument.
+  #   - Should return `true` if the rate limiter is active, otherwise `false`.
+  #
+  # @param global [Boolean] Whether the rate limiter applies globally across all sites. Defaults to `false`.
+  #   - Ignored if `klass` is provided.
+  #
+  # @param after [Class, nil] The rate limiter class after which the new rate limiter should be added.
+  #
+  # @param before [Class, nil] The rate limiter class before which the new rate limiter should be added.
+  #
+  # @example Adding a rate limiter that rate limits all requests from Googlebot in the same rate limit bucket.
+  #
+  #  add_request_rate_limiter(
+  #    identifier: :crawlers,
+  #    key: ->(request) { "crawlers" },
+  #    activate_when: ->(request) { request.user_agent&.include?("Googlebot") },
+  #  )
+  def add_request_rate_limiter(
```

Comment:
I don't think there will be a need for this API. I don't think we want to allow default rate limiters to be removed and I don't think there is a use case where a plugin needs to remove a custom rate limiter of another plugin.

Reply:
I was thinking maybe the admin might want to toggle on/off a rate limiter, but that's probably better expressed by a check for the site setting (or whatever) in the rate limiter itself rather than adding/removing it

Reply:
Yea I think the site setting approach will be better because we then get the UI for free.